generator/camera_info_collector.cpp

```diff
@@ -0,0 +1,340 @@
+
+      tmpCoef = MercatorBounds::DistanceOnEarth(p1, cameraProjOnSegment) /
+                MercatorBounds::DistanceOnEarth(p1, p2);
+    }
+
+    if (curMinDist < minDist && curMinDist < kMaxDistFromCameraToClosestSegmentMeters)
+    {
+      minDist = curMinDist;
+      bestFeatureId = ft.GetID().m_index;
+      segmentIdOfBestFeatureId = static_cast<uint32_t>(bestSegment.m_ind);
+      coef = tmpCoef;
+    }
+  };
+
+  dataSource.ForEachInRect(
+      updClosestFeatureCallback,
+      MercatorBounds::RectByCenterXYAndSizeInMeters(m_center, kSearchCameraRadiusMeters),
+      scales::GetUpperScale());
+
+  if (minDist != std::numeric_limits<double>::max())
+    m_ways.emplace_back(bestFeatureId, segmentIdOfBestFeatureId, coef);
+}
+
+boost::optional<std::pair<double, uint32_t>> CamerasInfoCollector::Camera::FindMyself(
+    uint32_t wayId, FrozenDataSource const & dataSource, MwmSet::MwmId const & mwmId) const
+{
+  double coef = 0;
+  bool isRoad = true;
+  uint32_t result = 0;
+
+  auto const readFeature = [&coef, &result, this, &isRoad, wayId](FeatureType & ft) {
+    bool found = false;
+    isRoad = routing::IsRoad(feature::TypesHolder(ft));
+    if (!isRoad)
+      return;
+
+    auto const forEachPoint = [&result, &found, this](m2::PointD const & pt) {
+      if (found)
+        return;
+
+      if (AlmostEqualAbs(m_center, pt, kEqualityEps))
+        found = true;
+      else
+        ++result;
+    };
+
+    ft.ForEachPoint(forEachPoint, scales::GetUpperScale());
+
+    CHECK(found, ("Cannot find camera point at the feature:", wayId));
+
+    // If point with number - N, is end of feature, we cannot say: segmentId = N,
+    // because number of segments is N - 1, so we say segmentId = N - 1, and coef = 1
+    // which means, that camera placed at the end of (N - 1)'th segment.
+    if (result == ft.GetPointsCount() - 1)
+    {
+      CHECK_NOT_EQUAL(result, 0, ("Feature consists of one point!"));
+      result--;
+      coef = 1;
+    }
+  };
+
+  FeatureID featureID(mwmId, wayId);
+  dataSource.ReadFeature(readFeature, featureID);
+
+  if (isRoad)
+    return boost::optional<std::pair<double, uint32_t>>({coef, result});
+  else
+    return {};
+}
+
+void CamerasInfoCollector::Camera::Serialize(FileWriter & writer,
+                                             uint32_t prevFeatureId) const
+{
+  CHECK(m_ways.size() == 1, ());
+
+  auto const & way = m_ways.back();
+
+  auto featureId = static_cast<uint32_t>(way.m_featureId);
+  featureId -= prevFeatureId;  // delta coding
+  WriteVarUint(writer, featureId);
+  WriteVarUint(writer, way.m_segmentId);
+
+  uint32_t const coef = DoubleToUint32(way.m_coef, 0, 1, 32);
+  WriteToSink(writer, coef);
+
+  WriteVarUint(writer, m_maxSpeedKmPH);
+
+  auto const direction = static_cast<uint8_t>(m_direction);
+  WriteVarUint(writer, direction);
+
+  // TODO (@gmoryes) add implementation of this feature
+  // Here will be a list of time conditions, for example:
```

Comment:
List of time conditions will be saved here. For example:

Reply:
done